cost_map_core/include/cost_map_core/operators/inflation.hpp

```diff
@@ -0,0 +1,196 @@
+/**
+ * @file /cost_map_core/include/cost_map_core/operators/inflation.hpp
+ */
+/*****************************************************************************
+** Ifdefs
+*****************************************************************************/
+
+#ifndef cost_map_core_INFLATION_HPP_
+#define cost_map_core_INFLATION_HPP_
+
+/*****************************************************************************
+** Includes
+*****************************************************************************/
+
+#include "../CostMap.hpp"
+#include <queue>
+
+/*****************************************************************************
+** Namespaces
+*****************************************************************************/
+
+namespace cost_map {
+
+/*****************************************************************************
+** Inflation Function
+*****************************************************************************/
+
+/**
+ * @brief Function which can compute costs for the inflation layer.
+ *
+ * Inherit from this to generate your own inflation
+ * functions.
+ */
+class InflationComputer {
+public:
+  InflationComputer() {};
+  virtual ~InflationComputer() {};
+
+  /** @brief  Given a distance, compute a cost.
+   *
+   * @param  distance The distance from an obstacle in cells
+   * @return A cost value for the distance
+   **/
+  virtual unsigned char operator()(const float &distance) const = 0;
+};
+
+/**
+ * @brief Function which can compute costs for the inflation layer.
+ *
+ * This class provides a default inflation function which works like the
+ * ROS inflation layer. Inherit from this to generate your own inflation
+ * functions.
+ */
+class ROSInflationComputer : public InflationComputer {
+public:
+  ROSInflationComputer(const float& inscribed_radius, const float& weight);
+
+  virtual ~ROSInflationComputer() {};
+
+  /** @brief  Given a distance, compute a cost.
+   *
+   * @param  distance The metric distance from an obstacle (distance = cell_distance*resolution)
+   * @return A cost value for the distance
+   **/
+  virtual unsigned char operator()(const float &distance) const;
+private:
+  float inscribed_radius_, weight_;
+};
+
+/*****************************************************************************
+** Inflation
+*****************************************************************************/
+
+class Inflate {
```

Comment:
There are several patterns for setting up a family of custom handlers in a class:

1. Additional functions in the `Inflate` class
2. Function callbacks with a specific signature
3. Polymorphic functors (as is the case here)
4. Templatised functors
5. Boost/c++11 style function bindings

1. has the drawback in that users can't go and design/experiment with their own functions off in their own package. 2. is probably the 'vanilla' option here. 3. has the advantage in that it can store state - you can dynamically configure it as you wish (candy), you can generate statistics while it executes and view at your leisure later (very useful) 4. same as 3, but without the need for polymorphism

Finally 5. is becoming the normal way of doing things, but it's still either boost or c++ dependent. So I avoid it when it is simpler to do so.